app/com/linkedin/drelephant/tuning/reenablement/TuneInReEnabler.java

```diff
@@ -0,0 +1,226 @@
+  }
+
+  private boolean isTuneInDisabledForSpecifiedDuration(TuningJobDefinition
+      tuningJobDefinition, long disabledDurationInMs) {
+    if (logger.isDebugEnabled()) {
+      logger.debug(String.format("Checking if TuneIn for job %s is disabled for more than duration in ms %s}",
+          tuningJobDefinition.job.id, disabledDurationInMs));
+    }
+    JobSuggestedParamSet latestSuggestedJobParamSet =
+        TuningHelper.getLatestSuggestedJobParamSet(tuningJobDefinition.job.id);
+    long tuneinDisabledEpochTime = latestSuggestedJobParamSet.updatedTs
+        .getTime();
+    long currentEpochTime = System.currentTimeMillis();
+    return (currentEpochTime - tuneinDisabledEpochTime) >= disabledDurationInMs;
+  }
+
+  private boolean isTuneInDisabledForSpecifiedIterations(TuningJobDefinition
+      tuningJobDefinition, int minExecutionCount) {
+    if (logger.isDebugEnabled()) {
+      logger.debug(String.format("Checking if job %s had min %s executions with TuneIn disabled",
+          tuningJobDefinition.job.id, minExecutionCount));
+    }
+    List<TuningJobExecutionParamSet> lastJobExecutionParamSets = TuningHelper.getLastNExecutionParamSets(
+        tuningJobDefinition.job.id, minExecutionCount);
+    if (lastJobExecutionParamSets == null || lastJobExecutionParamSets.size() != minExecutionCount) {
+      return false;
+    }
+    TuningJobExecutionParamSet executionWithTuningEnabled =
+        lastJobExecutionParamSets.stream().filter(el -> el.tuningEnabled).findFirst().orElse(null);
+    return (executionWithTuningEnabled == null);
+  }
+
+  private void processReEnablementForTuneIn(TuningJobDefinition tuningJobDefinition) {
+    logger.info("Initiating the process for re-enabling Auto-Tuning");
+    updateIterationCountForJob(tuningJobDefinition,
+        numberOfIterationAfterTuneinReEnable);
+    tuningJobDefinition.tuningDisabledReason = StringUtils.EMPTY;
+    updateJobSuggestedParamSetForReEnable(tuningJobDefinition);
+    tuningJobDefinition.tuningEnabled = true;
+    updateTuneinReEnableDetails(tuningJobDefinition.job);
+    tuningJobDefinition.update();
+  }
+
+  private void updateIterationCountForJob(TuningJobDefinition
+      tuningJobDefinition, int maxTuneinIterationAfterReEnable) {
+    int iterationCountWithSuggestedParamSet = TuningHelper
+        .getNumberOfValidSuggestedParamExecution( TuningHelper
+            .getTuningJobExecutionFromDefinition(tuningJobDefinition.job));
+    tuningJobDefinition.numberOfIterations = Math.max( tuningJobDefinition.
+        numberOfIterations, (iterationCountWithSuggestedParamSet +
+        maxTuneinIterationAfterReEnable));
+    logger.info(String.format("Updating the iterationCount to %s for jobDefintionId %s",
+        tuningJobDefinition.numberOfIterations, tuningJobDefinition.job.id));
+  }
+
+  private void updateJobSuggestedParamSetForReEnable(TuningJobDefinition tuningJobDefinition) {
+    unMarkCurrentBestParamSet(tuningJobDefinition.job.id);
```

Comment:
I think , there might be race condition . What will happen in case best parameters is un marked and there is no parameter in created state , and job have called getCurrentRunParameter

Reply:
In that case the default parameter will be SENT, which is the expected behaviour.